ql/src/test/org/apache/hadoop/hive/ql/cleanup/TestCleanupService.java

```diff
@@ -77,6 +83,54 @@ public void testEventualCleanupService_finishesCleanupBeforeExit() throws IOExce
     assertTrue(cleanupService.await(1, TimeUnit.MINUTES));
   }
 
+  /**
+   * Testing behaviour of ClearDanglingScratchDir service over local tmp files/dirs
+   * @throws Exception
+   */
+  @Test
+  public void localDanglingFilesCleaning() throws Exception {
+    HiveConf conf = new HiveConf();
+    conf.set("fs.default.name", "file:///");
+    FileSystem fs = FileSystem.get(conf);
+
+    // constants
+    String appId = "appId_" + System.currentTimeMillis();
+    String userName = System.getProperty("user.name");
+    String hdfs = "hdfs";
+    String inuse = "inuse.lck";
+    String l = File.separator;
+
+    // simulating hdfs dangling dir and its inuse.lck file
+    Path hdfsRootDir = new Path( HiveConf.getVar(conf, HiveConf.ConfVars.SCRATCHDIR) + l + userName + l + hdfs);
+    Path hdfsSessionDir = new Path(hdfsRootDir + l + userName + l + appId);
+    Path hdfsSessionLock = new Path(hdfsSessionDir + l + inuse);
+    fs.create(hdfsSessionLock);
+
+    // simulating local dangling files
+    String localTmpDir = HiveConf.getVar(conf, HiveConf.ConfVars.LOCALSCRATCHDIR);
+    Path localSessionDir = new Path(localTmpDir + l + appId);
+    Path localPipeOutFileRemove = new Path(localTmpDir + l
+            + appId + "-started-with-session-name.pipeout");
+    Path localPipeOutFileNotRemove = new Path(localTmpDir + l
+            + "not-started-with-session-name" + appId + ".pipeout");
+    fs.mkdirs(localSessionDir);
+    fs.create(localPipeOutFileRemove);
+    fs.create(localPipeOutFileNotRemove);
+
+    // running only the new method, the main service will be identifying which session files/dirs are dangling
+    ClearDanglingScratchDir clearDanglingScratchDirMain = new ClearDanglingScratchDir(false,
+            false, true, hdfsRootDir.toString(), conf);
+    clearDanglingScratchDirMain.run();
+
+    // should remove all except localPipeOutFileNotRemove, because it does not start with session name
+    Assert.assertFalse("Local session dir '" + localSessionDir
+            + "' still exists, should have been removed!", fs.exists(localSessionDir));
+    Assert.assertFalse("Local .pipeout file '" + localPipeOutFileRemove
+            + "' still exists, should have been removed!", fs.exists(localPipeOutFileRemove));
+    Assert.assertTrue("Local .pipeout file '" + localPipeOutFileNotRemove
+            + "' does not exist, should have not been removed!", fs.exists(localPipeOutFileNotRemove));
```

Comment:
You may also create a file which is:
1.  not the last in the listing [here](https://github.com/apache/hive/blob/9d17812f7d6958cf843b399ddfc6595f38137a11/ql/src/java/org/apache/hadoop/hive/ql/session/ClearDanglingScratchDir.java#L256), and check if the service still deletes every other files.
2. The service cannot delete it (either still open, or requires rights the service doesn't have)

You should check if all the other files are still deleted.

Reply:
So, now we have in total 3 files/dirs;
1. Local scratch dir belongs to session (starts with session name).
2. `.pipeout` file belongs to session (starts with session name).
3. `.pipeout` file does not belong to session (does not start with session name).

Before the service is executed;
![before](https://github.com/apache/hive/assets/73108963/62fac8df-4886-4c14-9743-4016a7ff1608)
After the service is executed;
![after](https://github.com/apache/hive/assets/73108963/478a5b34-3b5b-4dda-8542-1640a33fa6f8)

At the end, we check all three files, assert if file/dir #1 and #2 still exist, or #3 does not exist.
Is this sufficient?

Reply:
I meant the following: It would be nice to test if the delete fails for a particular file for whatever reason (the file is being used, different owner, etc), the loop does not terminate and the remaining files are still being cleaned up. I know the method now fulfills this (wit the canWrite check), but the test should also cover this edge case, so subsequent modifications won't break this without notice.

Reply:
Okay, I got your point.

So, I have added another file which its name starts with session name, that means the file will pass the filter [here](https://github.com/apache/hive/blob/9d17812f7d6958cf843b399ddfc6595f38137a11/ql/src/java/org/apache/hadoop/hive/ql/session/ClearDanglingScratchDir.java#L256).

But, as I have changed the file permission as read only, the file cannot be deleted because as you said we check whether the file is writable or not before delete;
![not_writeable](https://github.com/apache/hive/assets/73108963/e60bf22d-d5a3-49f2-9d4a-68ce494aea0f)

So, at the end of the test, there is one more assert now which makes sure that the file is not removed by the service.

Reply:
Excellent, this is exactly what I would like to have.